source/fab/repository.py

```diff
@@ -12,8 +12,9 @@
 """
 from abc import ABC, abstractmethod
 from pathlib import Path
-from subprocess import run
-from urllib.parse import urlparse
+from subprocess import PIPE, run, Popen
+import tarfile
+from urllib.parse import urlparse, urlunparse
```

Comment:
I think this could be another case for our style guidelines or conventions - I sometimes feel like the `from x import y` statements shouldn't be mixed with `import z` as it looks a bit strange/messy.  On the other hand - could you be explicit and do `from tarfile import ?` anyway?

Reply:
It would probably be a bad idea since what I am importing from `tarfile` is the function `open`. Obviously this would clash with the intrinsic open function.

I vascilate on how best to handle imports. For classes I like to import them seperately so I don't have to give full paths when referring to them. They tend to have reasonably unique names so collisions are not such a problem.

For global functions I prefer to import them through the module since they often have generic names like `open` or `run`.

Unfortunately it is hard to work out a good way to do both, see the subprocess line for an example. Ideally I would import just the Popen class on its own and the PIPE and run functions in their namespace.

Reply:
Ah yes, I see what you mean... I guess you could do `from tarfile import open as taropen` but that introduces your own naming decision into the mix, further obfuscating things.  For the `subprocess` case you could always do it on more than one line?  (have both `import subprocess` and `from subprocess import Popen`) - but again maybe that is just as bad (or worse!)

Reply:
I don't know. This is why I vascilate.